pkg/provisioners/gcppubsub/dispatcher/dispatcher/reconcile.go

```diff
@@ -0,0 +1,331 @@
+	}
+
+	if err = util.UpdateChannel(ctx, r.client, c); err != nil {
+		logging.FromContext(ctx).Info("Error updating Channel Status", zap.Error(err))
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, reconcileErr
+}
+
+// shouldReconcile determines if this Controller should control (and therefore reconcile) a given
+// ClusterChannelProvisioner. This Controller only handles gcp-pubsub Channels.
+func (r *reconciler) shouldReconcile(c *eventingv1alpha1.Channel) bool {
+	if c.Spec.Provisioner != nil {
+		return ccpcontroller.IsControlled(c.Spec.Provisioner)
+	}
+	return false
+}
+
+func (r *reconciler) reconcile(ctx context.Context, c *eventingv1alpha1.Channel) error {
+	// We are syncing all the subscribers on this Channel. Every subscriber will have a goroutine
+	// running in the background polling the GCP PubSub Subscription.
+
+	if c.DeletionTimestamp != nil {
+		// We use a finalizer to ensure we stop listening on the GCP PubSub Subscriptions.
+		r.stopAllSubscriptions(ctx, c)
+		util.RemoveFinalizer(c, finalizerName)
+		return nil
+	}
+
+	util.AddFinalizer(c, finalizerName)
+
+	err := r.syncSubscriptions(ctx, c)
+	return err
+}
+
+// key creates the first index into reconciler.subscriptions, based on the Channel's name.
+func key(c *eventingv1alpha1.Channel) channelName {
+	return types.NamespacedName{
+		Namespace: c.Namespace,
+		Name:      c.Name,
+	}
+}
+
+// subscriptionKey creates the second index into reconciler.subscriptions, based on the Subscriber's
+// name.
+func subscriptionKey(sub *v1alpha1.ChannelSubscriberSpec) subscriptionName {
+	return types.NamespacedName{
+		Namespace: sub.Ref.Namespace,
+		Name:      sub.Ref.Name,
+	}
+}
+
+// stopAllSubscriptions stops listening to all GCP PubSub Subscriptions for the given Channel.
+func (r *reconciler) stopAllSubscriptions(ctx context.Context, c *eventingv1alpha1.Channel) {
+	r.subscriptionsLock.Lock()
+	defer r.subscriptionsLock.Unlock()
+	r.stopAllSubscriptionsUnderLock(ctx, c)
+}
+
+// stopAllSubscriptionsUnderLock stops listening to all GCP PubSub Subscriptions for the given
+// Channel.
+// Note that it can only be called if reconciler.subscriptionsLock is held.
+func (r *reconciler) stopAllSubscriptionsUnderLock(ctx context.Context, c *eventingv1alpha1.Channel) {
+	channelKey := key(c)
+	if subscribers, present := r.subscriptions[channelKey]; present {
+		for _, subCancel := range subscribers {
+			subCancel()
+		}
+	}
+	delete(r.subscriptions, channelKey)
+}
+
+// syncSubscriptions ensures all subscribers of the Channel have a background Goroutine that is
+// polling the GCP PubSub Subscriptions representing it. It also removes listeners from Subscribers
+// that no longer exist.
+func (r *reconciler) syncSubscriptions(ctx context.Context, c *eventingv1alpha1.Channel) error {
+	r.subscriptionsLock.Lock()
+	defer r.subscriptionsLock.Unlock()
+
+	// We are going to manipulate subscribers, but don't want it written back, so make a throw away
+	// copy.
```

Comment:
Is this actually necessary? I don't see anything that writes to subscribers.

Reply:
I think an earlier version wrote to subscribers, but nothing does anymore. Removed.